src/org/rascalmpl/util/maven/SimpleResolver.java

```diff
@@ -90,9 +99,83 @@ private Path calculateJarPath(String groupId, String artifactId, String version,
+    // a different repo than the pom file. This should not pose a problem as pom files with the
+    // same version should be identical. If they are not, there is something seriously wrong with
+    // the versioning in one of the repos and we are helpless to fix that anyway.
+    private Metadata downloadArtifactMetadata(String groupId, String artifactId, String versionSpec)
+        throws UnresolvableModelException {
+
+        // TODO: deal with repository filters etc in settings.xml
+        String metadataUrl = String.format("/%s/%s/maven-metadata.xml", groupId.replace('.', '/'), artifactId);
+
+        MetadataXpp3Reader reader = new MetadataXpp3Reader();
+        for (var repoDownloader : availableRepostories) {
+            if (repoDownloader.getRepo().getLayout().equals("legacy")) {
+                // TODO: support legacy repo
+                continue;
+            }
+
+            String metadataSource = repoDownloader.read(metadataUrl);
+            if (metadataSource != null) {
+                try {
+                    return reader.read(new StringReader(metadataSource));
+                }
+                catch (IOException e) {
+                    throw new UnresolvableModelException("IOException trying to parse metadata", groupId, artifactId, versionSpec, e);
+                }
+                catch (XmlPullParserException e) {
+                    throw new UnresolvableModelException("Xml exception trying to parse metadata", groupId, artifactId, versionSpec, e);
+                }
+            }
+        }
+        throw new UnresolvableModelException("Could not download artifact metadata from available repositories", groupId,
+            artifactId, versionSpec);
+    }
+
+    private String findLatestMatchingVersion(Metadata metadata, String groupId, String artifactId, String versionRange) throws UnresolvableModelException {
+        try {
+            VersionRange range = VersionRange.createFromVersionSpec(versionRange);
+            Versioning versioning = metadata.getVersioning();
+            String latest = null;
+            DefaultArtifactVersion latestVersion = null;
+
+            for (String currentVersion : versioning.getVersions()) {
+                DefaultArtifactVersion candidate = new DefaultArtifactVersion(currentVersion);
+                if (range.containsVersion(candidate)) {
+                    if (latestVersion == null || candidate.compareTo(latestVersion) > 0) {
+                        latestVersion = candidate;
+                        latest = currentVersion;
+                    }
+                }
+            }
+
+            if (latest != null) {
+                return latest;
+            } else {
+                throw new UnresolvableModelException("No version found in range", groupId, artifactId, versionRange);
+            }
+        }
+        catch (InvalidVersionSpecificationException e) {
```

Comment:
shouldn't this be in the loop? else 1 wrong version string could break the loop over other versions? or is this only thrown by `VersionRange.createFromVersionSpec` and not by `new DefaultArtifactVersion()`?